src/providers/casbin-authorization-action.provider.ts

```diff
@@ -0,0 +1,86 @@
+import { Getter, inject, Provider } from '@loopback/core';
+import { IAuthUserWithPermissions } from '@sourceloop/core';
+import * as casbin from 'casbin';
+import * as fs from 'fs';
+import * as path from 'path';
+import { AuthorizationBindings } from '../keys';
+import { AuthorizationMetadata, CasbinAuthorizeFn, CasbinEnforcerConfigGetterFn } from '../types';
+const fsPromises = fs.promises;
+
+const DEFAULT_SCOPE = 'execute';
+
+export class CasbinAuthorizationProvider implements Provider<CasbinAuthorizeFn> {
+  constructor(
+    @inject.getter(AuthorizationBindings.METADATA)
+    private readonly getCasbinMetadata: Getter<AuthorizationMetadata>,
+    @inject.getter(AuthorizationBindings.CASBIN_ENFORCER_CONFIG_GETTER)
+    private readonly getCasbinEnforcerConfig: Getter<CasbinEnforcerConfigGetterFn>,
+  ) { }
+
+  value(): CasbinAuthorizeFn {
+    return (response, resource) => this.action(response, resource);
+  }
+
+  async action(user: IAuthUserWithPermissions, resource: string): Promise<boolean> {
+    let authDecision = false;
+    try {
+      const metadata: AuthorizationMetadata = await this.getCasbinMetadata();
+
+      const subject = this.getUserName(`${user.id}`);
+
+      const object = resource;
+
+      const action = metadata.permissions && metadata.permissions.length > 0 ? metadata.permissions[0] : DEFAULT_SCOPE;
+
+      const fn = await this.getCasbinEnforcerConfig();
+
+      const result = await fn(user, metadata.resource);
+
+      let enforcer: casbin.Enforcer;
+
+      if (metadata.isCasbinPolicy) {
```

Comment:
can u put some documentation of what's going around here ?

Reply:
done